utils/src/set.rs

```diff
@@ -392,6 +403,216 @@ impl<K, V> DoubleEndedIterator for OrderedAssociatedIntoIter<K, V> {
     }
 }
 
+/// An ordered, deduplicated slice of items each paired with some associated value, where values must be unique.
+///
+/// Like [`OrderedAssociated`], but enforces that values are unique across all keys. The contained
+/// [`Vec<(K, V)>`] is sealed after construction and cannot be modified. To unseal the inner
+/// [`Vec<(K, V)>`], use the [`Into<Vec<(K, V)>>`] impl.
+///
+/// Consumers that only need the ordered keys can treat an [`OrderedAssociatedUnique`] as an
+/// [`Ordered`] through deref coercions.
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct OrderedAssociatedUnique<K, V> {
```

Comment:
We create a distinct type here to make the invariant required clearer (if it was just `OrderedAssociated` with a unique constructor the user couldn't make any assumptions about its contents).